Spec: add types timstamp_ns and timestamptz_ns

jacobmarble opened the pull request:
Helps #8657

This change embodies this design doc:
https://docs.google.com/document/d/1bE1DcEGNzZAMiVJSZ0X1wElKLNkT9kRkk0hDlfkXzvU/edit

My IDE reformatted the Markdown tables edited. I'm happy to revert those if needed.

I also resorted the tables of types, so that date/time types land after other primitive types, and before nested types.

jacobmarble commented:
@Fokko - friendly reminder to review

rdblue commented:
@jacobmarble, I was talking with @danielcweeks a couple days ago about v3 additions and we came up with a good reason to add `timestamp_ms` and `timestamptz_ms`. Currently, type promotion is limited to very specific changes, but we think that we will need to open that up in v3. As long as we are open to promoting types, it makes sense to let people promote `long` to `timestamp` of some kind.

Now that we are adding timestamps with different units, this gives us a way to fix a long-standing problem: older data that used millisecond timestamps as longs couldn't be read as timestamps because we had no way to know the unit of the `long` (millis or micros). Now that we have separate types, we can use the type to carry that information, so that you can promote a `long` to `timestamp_ms` or `timestamp` and we know how to interpret the value.

What do you think? If you agree, then we should probably add `ms` types at the same time.

rdblue commented on format/spec.md:
```diff
@@ -167,30 +167,34 @@ A **`map`** is a collection of key-value pairs with a key type and a value type.
+| **`timestamp`**      | Timestamp, with microsecond precision, without timezone                  | [4], [6]                                         |
+| **`timestamptz`**    | Timestamp, with microsecond precision, with timezone                     | [4], [7]                                         |
+| **`timestamp_ns`**   | Timestamp, with nanosecond precision, without timezone                   | [5], [6]                                         |
+| **`timestamptz_ns`** | Timestamp, with nanosecond precision, with timezone                      | [5], [7]                                         |
```
We'll need to specify which types are allowed in different versions of the spec. All of the original types are allowed from v1 forward. The new types would need to be allowed from v3 forward.

jacobmarble replied:
I've added this to Appendix E.

jacobmarble commented:
> What do you think? If you agree, then we should probably add `ms` types at the same time.

Sounds good to me. I'll add milliseconds to this PR.

jacobmarble commented:
In the Iceberg community sync today, it was decided that we will not be adding millisecond timestamps after all.

Current status of this PR:
- millisecond timestamps removed
- note added to v3 Appendix
- rebased and squashed
- type promotion *not included*
  - someone with better context will follow up in a separate issue

@Fokko do you think we're ready to merge?

Fokko commented:
@jacobmarble Yes, this looks good to me. Thanks for joining the community sync. Could you avoid reformatting the table? History is especially important in the spec.

jacobmarble commented:
> @jacobmarble Yes, this looks good to me. Thanks for joining the community sync. Could you avoid reformatting the table? History is especially important in the spec.

Done.

Fokko commented on format/spec.md:
```diff
@@ -177,20 +177,24 @@ A **`map`** is a collection of key-value pairs with a key type and a value type.
-    - Timestamps _without time zone_ represent a date and time of day regardless of zone: the time value is independent of zone adjustments (`2017-11-16 17:10:34` is always retrieved as `2017-11-16 17:10:34`). Timestamp values are stored as a long that encodes microseconds from the unix epoch.
-3. Character strings must be stored as UTF-8 encoded byte arrays.
+1. `decimal(P,S)` scale (`S`) is fixed, and cannot be changed by schema evolution. Precision (`P`) can only be widened by schema evolution.
+2. `string` values must be stored as UTF-8 encoded byte arrays.
```
There are some discrepancies here. `[2]` is referenced in the time field, which doesn't seem applicable for the `time` type. Can you keep the changes to a minimum? 🙏

jacobmarble replied:
Yikes, you're right.

I've adjusted so that only `[2]` changes.

rdblue commented on format/spec.md:
```diff
@@ -862,10 +864,12 @@ Maps with non-string keys must use an array representation with the `map` logica
 |**`double`**|`double`||
 |**`decimal(P,S)`**|`{ "type": "fixed",`<br />&nbsp;&nbsp;`"size": minBytesRequired(P),`<br />&nbsp;&nbsp;`"logicalType": "decimal",`<br />&nbsp;&nbsp;`"precision": P,`<br />&nbsp;&nbsp;`"scale": S }`|Stored as fixed using the minimum number of bytes for the given precision.|
-|**`date`**|`{ "type": "int",`<br />&nbsp;&nbsp;`"logicalType": "date" }`|Stores days from the 1970-01-01.|
+|**`date`**|`{ "type": "int",`<br />&nbsp;&nbsp;`"logicalType": "date" }`|Stores days from 1970-01-01.|
```
No need to change this line.

rdblue commented on format/spec.md:
```diff
@@ -874,6 +878,11 @@ Maps with non-string keys must use an array representation with the `map` logica
+Notes:
+
+1. Avro type annotation `adjust-to-utc` is an Iceberg convention; default value is `false` if not present.
+2. Avro logical type `timestamp-nanos` is an Iceberg convention; the Avro specification does not define this type.
```
Is this necessary? Shouldn't we just go add this type to Avro, or use an Avro version?

Fokko replied:
In the Google doc I already suggested adding this to the Avro spec: https://docs.google.com/document/d/1bE1DcEGNzZAMiVJSZ0X1wElKLNkT9kRkk0hDlfkXzvU/edit?disco=AAAA5YbF34Y

Let me sent out an email on the Avro devlist

jacobmarble replied:
@Fokko please drop a link to Avro devlist post when you have a moment. The mailing list link on [the Avro community page](https://avro.apache.org/community/) is just an email address, can't find the web UI.

Fokko replied:
Ah, we should update that page, that's not helping anyone. You can easily see the devlist on ponymail: https://lists.apache.org/list.html?dev@avro.apache.org

On the bottom-left you can subscribe to the devlist. You can find my message here: https://lists.apache.org/thread/s985tc2m6bj5lvck017q06zkt3dg1f31 or the PR https://github.com/apache/avro/pull/2554. Feel free to chime in anywhere.

jacobmarble replied:
@rdblue @Fokko I agree that we should ask this type be added to Avro.

It will be months before this change is added to Avro, then supported by dependent projects. Do you propose that:
- We wait for the Avro spec to change before modifying the Iceberg spec?
- Adjust the Iceberg spec language to reflect the intention to switch to the Avro type when it becomes available?

Fokko replied:
Since most Iceberg implementations have their own Avro readers, I think we're fine with running this in parallel. I don't see any huge issues on the Avro side (maybe some discussion around the physical type), but I would not consider that blocking.

jacobmarble replied:
@Fokko I understand; do I need to change this PR, then?

Fokko replied:
It depends on what comes out of the Avro PR, but I think we're good for now.

rdblue commented on format/spec.md:
```diff
@@ -948,6 +961,7 @@ Lists must use the [3-level representation](https://github.com/apache/parquet-fo
 Notes:
 
 1. ORC's [TimestampColumnVector](https://orc.apache.org/api/hive-storage-api/org/apache/hadoop/hive/ql/exec/vector/TimestampColumnVector.html) consists of a time field (milliseconds since epoch) and a nanos field (nanoseconds within the second). Hence the milliseconds within the second are reported twice; once in the time field and again in the nanos field. The read adapter should only use millis ...
+2. ORC `timestamp` and `timestamp_instant` values store nanosecond precision. Iceberg ORC writers for Iceberg types `timestamp` and `timestamptz` truncate nanoseconds to microseconds.
```
I think that this should state that Iceberg ORC writers MUST truncate nanoseconds to microseconds. Otherwise someone may store nanosecond values, which would break equality and have implications when casting from timestamp to timestamp_ns.

jacobmarble replied:
Done. PTAL https://github.com/apache/iceberg/pull/8683/commits/150b6a9e19a45c0741c3d1cd563e92708983bf2d

rdblue commented on format/spec.md:
```diff
@@ -971,8 +985,10 @@ The 32-bit hash implementation is 32-bit Murmur3 hash, x86 variant, seeded with
+| **`timestamp`**    | `hashLong(microsecsFromUnixEpoch(v))`     | `2017-11-16T22:31:08` ￫ `-2047944441`<br />`2017-11-16T22:31:08.000001` ￫ `-1207196810` |
+| **`timestamptz`**  | `hashLong(microsecsFromUnixEpoch(v))`     | `2017-11-16T14:31:08-08:00` ￫ `-2047944441`<br />`2017-11-16T14:31:08.000001-08:00` ￫ `-1207196810` |
+| **`timestamp_ns`** | `hashLong(nanosecsFromUnixEpoch(v))`      | `2017-11-16T22:31:08` ￫ `-737750069`<br />`2017-11-16T22:31:08.000001` ￫ `-976603392`<br />`2017-11-16T22:31:08.000000001` ￫ `-160215926` |
+| **`timestamptz_ns`** | `hashLong(nanosecsFromUnixEpoch(v))`    | `2017-11-16T14:31:08-08:00` ￫ `-737750069`<br />`2017-11-16T14:31:08.000001-08:00` ￫ `-976603392`<br />`2017-11-16T14:31:08.000000001-08:00` ￫ `-160215926` |
```
Can you please add tests for these cases? We need to make sure that the actual code produces the given values.

jacobmarble replied:
Yes. I did generate these values with code tests, which I did not add to this PR. This change is just a spec change, following [this comment](https://github.com/apache/iceberg/pull/8658#issuecomment-1740908740).

Would you like some code to change in this PR? If so, I could remove the milliseconds details in #8658 then add those commits to this PR.

jacobmarble replied:
I've added the unit tests to #8658 : https://github.com/apache/iceberg/pull/8658/commits/bc50195cc0f78d42cb18b6fa8021f882066c694d

jacobmarble commented:
@rdblue @Fokko what concerns remain regarding this change?

rdblue commented on format/spec.md:
```diff
@@ -187,10 +189,11 @@ A **`map`** is a collection of key-value pairs with a key type and a value type.
+    - Timestamp values _with time zone_ represent a point in time: values are stored as UTC and do not retain a source time zone (`2017-11-16 17:10:34 PST` is stored/retrieved as `2017-11-17 01:10:34 UTC` and these values are considered identical).
+    - Timestamp values _without time zone_ represent a date and time of day regardless of zone: the time value is independent of zone adjustments (`2017-11-16 17:10:34` is always retrieved as `2017-11-16 17:10:34`).
 3. Character strings must be stored as UTF-8 encoded byte arrays.
+4. `timestamp_ns` and `timstamptz_ns` are only supported in v3 tables.
```
I think that this needs to be a column in the table rather than a footnote. It just isn't very obvious that you can't use the ns types in v2 otherwise. That's my only concern with this PR.

jacobmarble replied:
Fixed.

Fokko replied:
I like Ryan's suggestion of having a full new column. We can expect more types to be added so that a column would make sense to me, instead of adding it to the `requirements` column.

jacobmarble replied:
Ah, I missed that detail. How's this?

rdblue commented:
@jacobmarble sorry for the delay, I was travelling. I think we just need to call out more clearly that the new types are v3 only and can't be used in v2 without breaking forward compatibility. A footnote isn't quite enough.

findepi commented:
> Now that we have separate types, we can use the type to carry that information, so that you can promote a `long` to `timestamp_ms` or `timestamp` and we know how to interpret the value.

That's true as long as there is only type promotion only once.

What should happen when `timestamp_ms` is promoted to `timestamp`?
If it wasn't a `long` before, such promotion should be permitted.
If it was a `long` before... how do we handle this case?

jacobmarble commented:
> > Now that we have separate types, we can use the type to carry that information, so that you can promote a `long` to `timestamp_ms` or `timestamp` and we know how to interpret the value.
> 
> That's true as long as there is only type promotion only once.
> 
> What should happen when `timestamp_ms` is promoted to `timestamp`? If it wasn't a `long` before, such promotion should be permitted. If it was a `long` before... how do we handle this case?

FWIW, the millisecond timestamp types (`timestamp_ms`, `timestamptz_ms`) have been removed from this change.

findepi commented:
@jacobmarble indeed, thanks!
do we expect any type promotions being allowed around the new types being added here?

jacobmarble commented:
> @jacobmarble indeed, thanks! do we expect any type promotions being allowed around the new types being added here?

@findepi when we discussed in the last community meeting, that was discussed. We agreed to scope this PR, and #8657 to just adding the type, and let type promotion be a separate issue.

rdblue commented:
@findepi, for type promotion we decided to always interpret a promoted long as a millisecond value and the convert to either micros or nanos depending on the Iceberg type. Since we don't have a way to know what a long represents, there are problems no matter what we do and that's the best way to get predictable behavior. Otherwise, type promotion (i.e. timestamp_ms to a higher precision value) could corrupt data values that were previously working.

rdblue commented on format/spec.md:
```diff
@@ -177,8 +177,10 @@ A **`map`** is a collection of key-value pairs with a key type and a value type.
+| **`timestamp`**    | Timestamp, microsecond precision, without timezone                       | [2]                                              |
+| **`timestamptz`**  | Timestamp, microsecond precision, with timezone                          | [2]                                              |
+| **`timestamp_ns`** | Timestamp, nanosecond precision, without timezone                        | [2]                                              | [v3](#version-3) |
+| **`timestamptz_ns`** | Timestamp, nanosecond precision, with timezone                         | [2]                                              | [v3](#version-3) |
```
I don't think this is clear enough. I think the version column should come first and be "Valid from" or similar. "Iceberg version" could be confused with the library version and we need to make it clear what this version means: that you are not allowed to use it in tables until that version.

jacobmarble replied:
PTAL. I could also fill in the other rows with v1.

rdblue commented:
@jacobmarble, I updated the table to be more clear about when types can be used and merge this. Thanks!
